Chapter two, first draft for review.

NicholasWoodIMG opened the pull request.

rsahita commented on specification/images/img_ch1_reference-model.png:Zone.Identifier:
is there a short link for this drawing? cannot follow this link.

NicholasWoodIMG replied:
I don't actually know where these "Zone.Identifier" files come from... They randomly appear in my local git working area. Possibly the Adoc preview in VS Code... I have just ignored them for now, we can clean up the images directory once all the merges have completed.

rsahita commented on specification/images/img_ch2_reference-model.png:
this image is very weird for me - should we show priv levels for the workloads? should the isa spec blocks go under the TSS since they are all usable by some priv levels - do the levels of the isa match the workload levels?

the system hardware blocks are also hard to follow - what does trusted security device - isnt a RoT one of those?

I will come back and add more comments after reading the chapter.

NicholasWoodIMG replied:
The intent of this picture is just to define generic functional components of any secure system without getting bogged down in all the combinations and permutations of actual Risc-V architectural building blocks. So I stripped it right back.

Maybe I took it too far... :-) To be honest, I'm not happy with this section myself. 

Looking at it again:

- Do you have some time to sit down and work through terminology and intent? Probably quicker than a text review.
- We might restructure the whole thing as "security guarantees" instead, with high-level guarantees. I can make a proposal for that separately that we can discuss. Basically the RME approach instead of the PSA approach.

NicholasWoodIMG replied:
I have completely re-written the reference model. I am sure it will need a tweak here and there but I am much happier with this version. Hope you agree.

I have pushed it to this PR, hope you can see the update.

rsahita commented on specification/src/chapter2.adoc:
```suggestion
| Current RISC-V mitigations 
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
| Planned RISC-V mitigations
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
a| * RISC-V (V)U/(V)S/S(H)/M privilege levels
```

NicholasWoodIMG replied:
I would prefer to just keep "Risc-V privilege levels". The full definition of those is later, I don't think we need to keep enumerating that everywhere.

rsahita commented on specification/src/chapter2.adoc:
```suggestion
* RISC-V supervisor domains (MTT)
```

NicholasWoodIMG replied:
This is a catch-all for any Risc-V isolation mechanism. These are enumerated later. It's not just supervisor domains.

rsahita commented on specification/src/chapter2.adoc:
```suggestion
* RISC-V hardware virtualization (H extension)
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
| Known (documented) attacks except Spectre v1 are specific to particular micro-architectures, and RISC-V systems are not expected to be vulnerable to those. This is an evolving area of research. +
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
a| * RISC-V (V)U/(V)S/S(H)/M privilege levels
```

NicholasWoodIMG replied:
I would prefer to just keep "Risc-V privilege levels". The full definition of those is later, I don't think we need to keep enumerating that everywhere.

rsahita commented on specification/src/chapter2.adoc:
```suggestion
* RISC-V Supervisor domains (MTT)
```

NicholasWoodIMG replied:
This is a catch-all for any Risc-V isolation mechanism. These are enumerated later. It's not just supervisor domains, and not just MTT.

Do you agree?

rsahita commented on specification/src/chapter2.adoc:
```suggestion
a| * Shadow stacks (Zicfiss)
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
* Landing pads (Zicfilp)
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
| RISC-V recommendations 
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
* Data Independent Execution Latency (Zkt, Zvkt)
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
In some cases RISC-V non-ISA specifications can provide guidance or protocols. This is discussed more in use case examples later in this specification.
```

rsahita commented on specification/src/chapter2.adoc:
```suggestion
Additional application specific provisioning stages can take place in this state - for example network onboarding and device activation, TSS/App/Device attestation or user identity management. This is out of scope of this specification.
```

rsahita commented on specification/src/chapter2.adoc:
should verification be optional (may) since some platforms may find measured boot with attestation sufficient?

NicholasWoodIMG replied:
I have updated the "Authorized software" section. Please review.

rsahita commented on specification/src/chapter2.adoc:
this will be limiting capabilities that can securely update a layered TCB component without restarting dependent software component. (it is expected in these cases, the dependent SW components explicitly opt-in to updates, and the attestation reports that opt-in apriori along with the creation security version and the current security version of the TCB.

NicholasWoodIMG replied:
This is the catch-all. Live-updates are covered just below. Are you ok with that?

NicholasWoodIMG replied:
I have updated the "Authorized software" section above to prepare the ground. Please review.

Will tweak the update section tomorrow.

NicholasWoodIMG replied:
Pushed an update to the system update section as well. Please review.

rsahita commented on specification/src/chapter2.adoc:
most systems now want to update without a full reset.

NicholasWoodIMG replied:
In server land maybe :-) 

Ok, I will review this section and suggest an update. We are basically talking about either local boot rooted in immutable boot code, where at least the trusted security services must be measure and verified; or host attestation where system boot only is measured, but the the host must be attested before launching hosted services.

NicholasWoodIMG replied:
I have updated the "Authorized software" section above to prepare the ground. Please review.

Will tweak the update section tomorrow.

ved-rivos replied:
Also in other lands like industrial control systems that run mission critical processes.

NicholasWoodIMG replied:
Pushed an update to the system update section as well. Please review.

rsahita commented on specification/src/chapter2.adoc:
```suggestion
Use cases later in this specification provide examples of RISC-V isolation models.
```

rsahita reviewed (changes requested):
some comments and feedback added.
thanks

ved-rivos commented:
This was sent previously in an email. Now that we have a PR, attaching the comments here:
1. In reference model, should "System hardware" be "System
    hardware/firmware"?
2. Logic:Unauthorized-access:
     - does "RISC-V isoaltion" include PMP, virtual memory, MTT?
     - Virtualization - different than privilege level?
3. Logical:Interface abuse: The high assurance crypto. seems
    more appropriate for "Analysis of physical leakage".
4. "Subverting supply chains" -> should it include attestation?
    Should it also assume physical - e.g., stealing unfused parts, etc.
5. Authorized sw - should there be a statement about authorities
    for verified launch?
6. "An update of a softwae component SHOULD restart
     dependent...".
     For servers it is becoming increasingly important to be able
     to support live updates without restart to avoid violating
     service level objectives. Should this normativbe requirement
     also include that such restart is not required for live
     updateable systems that have a contract as mentioned later?
7. "An update of trusted...MUST only take effect following a
     system reset"
    This may be too restritive. The udpate process may provide an
    assurance that the update can only be a better TCB e.g.,
    something with a higher security version number. Mandating a
    restart seems very restrictive. Should this be replaced by
    key properties to achieve?
8. "Sealed data MUST only be acc..." - nit - Should it say
    "if it belongs to that component"?.
9. s/Risc V/RISC-V

NicholasWoodIMG commented:
@ved-rivos 

>     1. In reference model, should "System hardware" be "System
>        hardware/firmware"?

I have rewritten the whole "reference model" section. Please review the update (should be visible from the PR).

>     2. Logic:Unauthorized-access:
>        
>        * does "RISC-V isoaltion" include PMP, virtual memory, MTT?
>        * Virtualization - different than privilege level?

Yes. I added examples.

Virtualization - in principle just another privilege level, but I think it is important enough to call out. Added MMU as example there as well.

> 
>     3. Logical:Interface abuse: The high assurance crypto. seems
>        more appropriate for "Analysis of physical leakage".

Agree, but Markku seemed to differ. I think he is more concerned about outcome than method. So even if HAC does not actually prevent interface abuse, it could perhaps help reduce the consequences. I'm not sure if this is actually true - does not matter if a key is wrapped or not as the attacker can still use it (on the same system at least) to decrypt data. But I'm happy not picking that fight in this document.

> 
>     4. "Subverting supply chains" -> should it include attestation?
>        Should it also assume physical - e.g., stealing unfused parts, etc.
> 

Added "attestation processes, development processes (for example, unfused development hardware or debug authorizations)"

>     5. Authorized sw - should there be a statement about authorities
>        for verified launch?

Yes. I have updated this section from other comments as well. See latest version (should be visible in the PR).

>     6. "An update of a softwae component SHOULD restart
>        dependent...".
>        For servers it is becoming increasingly important to be able
>        to support live updates without restart to avoid violating
>        service level objectives. Should this normativbe requirement
>        also include that such restart is not required for live
>        updateable systems that have a contract as mentioned later?

Yes, this section has been updated. See latest draft in the PR.

>     7. "An update of trusted...MUST only take effect following a
>        system reset"
>        This may be too restritive. The udpate process may provide an
>        assurance that the update can only be a better TCB e.g.,
>        something with a higher security version number. Mandating a
>        restart seems very restrictive. Should this be replaced by
>        key properties to achieve?

Yes, this section has been updated. See latest draft in the PR.
 
>     8. "Sealed data MUST only be acc..." - nit - Should it say
>        "if it belongs to that component"?.

Ok.

>     9. s/Risc V/RISC-V

Done.
